File: packages/depz-sensor-sdk-cpp/src/sr04.cpp
#include "depz/sr04.hpp"

#include "depz/detail/byteio.hpp"

namespace depz {

using detail::rd_u16;
using detail::rd_u32;
using detail::rd_u64;
using detail::require;
using detail::u8;
using detail::wr_le;

Sr04Data Sr04Data::unpack(byte_span p) {
    require(p, 11, "Sr04Data payload too short");
    return Sr04Data{u8(p[0]), rd_u64(p, 1), rd_u16(p, 9)};
}

bytes pack_sample_period(std::uint32_t period_us) {
    bytes out;
    wr_le(out, period_us, 4);
    return out;
}

std::uint32_t unpack_sample_period(byte_span p) {
    require(p, 4, "sample_period payload too short");
    return rd_u32(p, 0);
}

bytes pack_echo_decay(std::uint16_t decay_us) {
    bytes out;
    wr_le(out, decay_us, 2);
    return out;
}

std::uint16_t unpack_echo_decay(byte_span p) {
    require(p, 2, "echo_decay payload too short");
    return rd_u16(p, 0);
}

std::optional<double> distance_mm_from_echo(std::uint16_t echo_time_us,
                                            std::optional<double> air_temp_c) {
    if (echo_time_us == ECHO_TIMEOUT) return std::nullopt;
    const double c_m_s = air_temp_c ? (331.3 + 0.606 * *air_temp_c) : 343.0;
    return echo_time_us * c_m_s / 2000.0;
}

}  // namespace depz

File: packages/depz-sensor-sdk-csharp/src/Depz.Sensor/Protocol/Sr04.cs
using System.Buffers.Binary;

namespace Depz.Sensor.Protocol;

/// <summary>SR04 command opcodes (contracts/03_SENSOR_SR04.md).</summary>
public enum Sr04Cmd
{
    GetSamplePeriod = 0x32,
    SetSamplePeriod = 0x33,
    GetEchoDecay = 0x34,
    SetEchoDecay = 0x35,
    MeasureOnce = 0x36,
    StartMeasurementLoop = 0x37,
    StopMeasurementLoop = 0x38,
}

/// <summary>SR04 report opcodes.</summary>
public enum Sr04Rpt
{
    Data = 0x91,
    SamplePeriod = 0x92,
    EchoDecay = 0x93,
}

/// <summary>SR04 measurement sample: source cmd (0x36 single / 0x37 loop), timestamp, echo time.</summary>
public sealed record Sr04Data(int SourceCmd, ulong TimestampUs, int EchoTimeUs)
{
    public static Sr04Data Unpack(ReadOnlySpan<byte> payload) => new(
        payload[0],
        BinaryPrimitives.ReadUInt64LittleEndian(payload[1..]),
        BinaryPrimitives.ReadUInt16LittleEndian(payload[9..]));
}

public static class Sr04
{
    /// <summary>echo_time_us sentinel: no echo received.</summary>
    public const int EchoTimeout = 0xFFFF;

    public const int SamplePeriodDefaultUs = 50_000;
    public const int EchoDecayDefaultUs = 5_000;
    public const int EchoDecayMinUs = 4_000;
    public const int EchoDecayMaxUs = 65_000;

    public static byte[] PackSamplePeriod(uint periodUs)
    {
        var buf = new byte[4];
        BinaryPrimitives.WriteUInt32LittleEndian(buf, periodUs);
        return buf;
    }

    public static uint UnpackSamplePeriod(ReadOnlySpan<byte> payload) =>
        BinaryPrimitives.ReadUInt32LittleEndian(payload);

    public static byte[] PackEchoDecay(ushort decayUs)
    {
        var buf = new byte[2];
        BinaryPrimitives.WriteUInt16LittleEndian(buf, decayUs);
        return buf;
    }

    public static ushort UnpackEchoDecay(ReadOnlySpan<byte> payload) =>
        BinaryPrimitives.ReadUInt16LittleEndian(payload);

    /// <summary>
    /// Round-trip echo time → distance in mm; null for the timeout sentinel.
    /// Default speed of sound 343 m/s; with <paramref name="airTempC"/> uses
    /// c = 331.3 + 0.606·T (m/s).
    /// </summary>
    public static double? DistanceMmFromEcho(int echoTimeUs, double? airTempC = null)
    {
        if (echoTimeUs == EchoTimeout)
            return null;
        double cMs = airTempC is null ? 343.0 : 331.3 + 0.606 * airTempC.Value;
        return echoTimeUs * cMs / 2000.0;
    }
}
